From 6ce8a31b0b87f9dc7ce94c4cba81074510b8c384 Mon Sep 17 00:00:00 2001
From: Akihiro Sagawa <sagawa.aki@gmail.com>
Date: Wed, 6 Mar 2024 23:29:31 +0900
Subject: [PATCH] dsound: Initialize MTA in IDirectSound::Initialize().

RE:D Cherish! (Trial ver) depends on this behavior.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=53613
---
 dlls/dsound/dsound.c         | 3 +++
 dlls/dsound/dsound_private.h | 1 +
 dlls/dsound/tests/dsound.c   | 4 ----
 dlls/dsound/tests/dsound8.c  | 4 ----
 4 files changed, 4 insertions(+), 8 deletions(-)

diff --git a/dlls/dsound/dsound.c b/dlls/dsound/dsound.c
index 57fd6c1a6517..7629810db505 100644
--- a/dlls/dsound/dsound.c
+++ b/dlls/dsound/dsound.c
@@ -207,6 +207,8 @@ static ULONG DirectSoundDevice_Release(DirectSoundDevice * device)
             WaitForSingleObject(device->thread, INFINITE);
             CloseHandle(device->thread);
         }
+        if (device->mta_cookie)
+            CoDecrementMTAUsage(device->mta_cookie);
 
         EnterCriticalSection(&DSOUND_renderers_lock);
         list_remove(&device->entry);
@@ -332,6 +334,7 @@ static HRESULT DirectSoundDevice_Initialize(DirectSoundDevice ** ppDevice, LPCGU
         WARN("DSOUND_ReopenDevice failed: %08lx\n", hr);
         return hr;
     }
+    CoIncrementMTAUsage(&device->mta_cookie);
 
     ZeroMemory(&device->drvcaps, sizeof(device->drvcaps));
 
diff --git a/dlls/dsound/dsound_private.h b/dlls/dsound/dsound_private.h
index 34498125317c..28bba6cd64ec 100644
--- a/dlls/dsound/dsound_private.h
+++ b/dlls/dsound/dsound_private.h
@@ -90,6 +90,7 @@ struct DirectSoundDevice
     int                         lfe_channel;
     float *tmp_buffer, *cp_buffer;
     DWORD                       tmp_buffer_len, cp_buffer_len;
+    CO_MTA_USAGE_COOKIE         mta_cookie;
 
     DSVOLUMEPAN                 volpan;
 
